### PR TITLE
fix: Simplify Fields to Param in Tracker Export [ DHIS2-14752 ]

### PR DESCRIPTION
**Pull request by lucaCambi77**
https://dhis2.atlassian.net/browse/DHIS2-14752
 
Finalize the work started with https://dhis2.atlassian.net/browse/DHIS2-14783

We want to simplify the way we map the request fields to the following Params classes :

- EnrollmentParams
- EventParams
- RelationshipParams
- TrackedEntityInstanceParams

These classes include or exclude fields in the payload response by triggering or not the query to fetch them.
So if we request `/tracker/trackedEntitties?fields=enrollments` we want to include the enrollments in the response, thus querying the DB for them.

Work done:

**1. Params Abstraction and Builder**

This PR adds the `AbstractParams` abstraction with a simple set of Enum fields named `Param`. So, we remove the boolean properties we had before.
This is how, for example, `TrackedEntityInstanceParams` will now look in debugging:
Screenshot_20230225_103255

The Enum comprises a prefix (if any) and a field which together map to the actual field name.
For example, for `TrackedEntityInstanceParams` we can have the Enum:

-  `ENROLLMENTS_EVENTS_RELATIONSHIPS( ENROLLMENTS, EVENTS_RELATIONSHIPS.getFieldPath() )`

composed by the prefix ENROLLMENTS and the field path for EVENTS_RELATIONSHIPS. This will create the field `enrollments.events.relationships` which maps to the actual field path.
When it comes to nested params, previously that was a property of the outer param class. Now we'll create them only if required and this is done just by removing the prefix, so 
- `enrollments.events.relationships` in TrackedEntityInstanceParams will become `events.relationships` in EnrollmentParams by removing the prefix.

To create a Params implementation object we can now use a builder. The builder will update the fields Enum list by creating new immutable objects. Furthermore, we replace the TRUE and FALSE initialization because it was not clear:
- TRUE replaced with all() method in the builder
- FALSE replaced with empty() method in the builder

For example, if we want to create EventParams with no fields, we'll now do `EventParams.builder().empty().build()`

**2. Remove dataSyncronizationQuery**

With this PR we also remove `dataSyncronizationQuery` from the Params because only in use in the `TrackerSynchronization` class. The `synchronizationQuery` is already present in the `TrackedEntityInstanceQueryParams` so we can use that instead and move away the logic from the Params classes
Finally, we remove the `dataSyncronizationQuery` from the old tracker endpoints because apparently it is not used.
Thus, it will remain only in the `TrackerSynchronization` class

We could argue at this point that using a list of enums instead of booleans is not good for performances. However, at least now we have an abstraction for the Params classes, clean code, and maintainability

**Open points / Ideas**

- We have different behaviors for the entities' endpoints. For example, in the `TrackerEventsExportController` we don't do the fields to params conversion, so we should adapt to the others or find a common way to do that
- Should we use a cache for these params classes? We map the fields given the table definition using reflection, so this is not going to change during runtime. We can think of a way to safely reuse given a fields request
- Now that we assign single responsibilities to every params class, we can give the possibility to users to define the inclusion of fields such as deleted for every entity. So instead of only one `includeDeleted` in the tracker/trackedEntities request, we can define one for enrollments and events. Not sure this can help or make sense

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/Param.java [hunk 1]
```diff
@@ -27,24 +27,57 @@
...
 {
-    public static final EnrollmentEventsParams TRUE = new EnrollmentEventsParams( true, EventParams.TRUE );
+    ATTRIBUTES( null, "attributes" ),
+    RELATIONSHIPS( null, "relationships" ),
+    PROGRAM_OWNERS( null, "programOwners" ),
+    DELETED( null, "deleted" ),
+    ENROLLMENTS( null, "enrollments" ),
+    EVENTS( null, "events" ),
+    EVENTS_RELATIONSHIPS( EVENTS, RELATIONSHIPS.getFieldPath() ),
+    ENROLLMENTS_EVENTS( ENROLLMENTS, EVENTS.getFieldPath() ),
```
[the remark this answers is not in the document]
Comment (lucaCambi77) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/Param.java:45]: yes i thought the same but there are very few params so it was going to be redundant in most cases

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/trackedentity/AbstractTrackedEntityInstanceService.java [hunk 2]
```diff
@@ -1678,12 +1683,9 @@ private TrackedEntityInstance getTei( org.hisp.dhis.trackedentity.TrackedEntityI
 
         }
 
-        Set<TrackedEntityAttribute> readableAttributesCopy = filterOutSkipSyncAttributesIfApplies( params,
```
[the remark this answers is not in the document]
Comment (lucaCambi77) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/trackedentity/AbstractTrackedEntityInstanceService.java:1681]: Sync (or better the `dataSyncronizationQuery` thing  ) apparently is not in use in the old tracker so I've removed it

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tracker/export/fieldsmapper/FieldsParamMapper.java [hunk 3]
```diff
@@ -31,27 +31,26 @@
 import java.util.List;
 import java.util.Map;
 
+import org.hisp.dhis.dxf2.events.AbstractParams;
```
[the remark this answers is not in the document]
Comment (enricocolasante) [reply dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tracker/export/fieldsmapper/FieldsParamMapper.java:34]: I really have some concerns on adding dependencies from `dxf2` here.
I really think we should have a deeper look at the first open point before we go through with this.
I would really like that when we finally decide to remove the old tracker we can just remove one package (obviously this is not gonna happen), or anyway avoid the need to go through the code and remove small bits here and there.

So the question here is really if it make sense to have all common parts in the `dhis-api` module or if it even make sense to duplicate some code, leaving the old one as it is and move the new one closer to the web

**Comment by stian-sandvold** [comment]
I think overall the idea of the refactor is sensible. It simplifies the code related to including sub-models of tracker entities. I kind of agree with @teleivo regarding having each entity responsible for their own fields, so it's clearer what is available where.

However, as mentioned, this now resides in the dxf2 package, which is the old tracker. The plan is to remove most, if not all, of this code next version (2.41; Since we don't have time now). That means immediately this refactor will be gone.

If we agree, meaning the rest of the tracker team as well, that this is a sensible refactor for tracker ( Personally I agree, but I'm not the one most actively working on this code any more ), then we should do this exclusively for the new tracker, not the old (at all). Here's why:
- If we do it for old tracker, that means we will have it around for a single version, then remove it. It's kind of wasted effort.
- If there are bugs with this code, or we need to change it, that means backporting a fix for old tracker now requires two different sets of code to fix it.
- By doing this only for the new tracker, that means we can detach further from the old tracker, when it comes down to removing (and cleaning up dependant code) the old tracker code.


So to summarize, let's move this to the new tracker, not keep it in the old tracker code, and let's get feedback of the team, who will be working on this code :)

What do you think?

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tracker/export/fieldsmapper/FieldsParamMapper.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mortenoh) [reply dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tracker/export/fieldsmapper/FieldsParamMapper.java:34]: Please also know that we are hoping to remove the `dxf2` module in the future. When the old tracker importer is gone, we want this module to only contain metadata importer code.

[the remark this answers is not in the document]
Comment (lucaCambi77) [reply dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tracker/export/fieldsmapper/FieldsParamMapper.java:32]: @teleivo @enricocolasante @stian-sandvold I see there is some confusion here. The `Params` classes were already in the dxf2 package so we are not adding dependencies here. The `Params` are used by the export endpoints using services all the way through the `dxf2` package. So I am not sure what is the best solution but duplicating the code or moving to the web part what does it mean? These are classes used in the services so in my opinion if we just move this to the `dhis-api` then later on when we get rid of the old tracker, we can move this stuff into the tracker module. If we duplicate code now we just have two things to maintain

[the remark this answers is not in the document]
Comment (stian-sandvold) [reply dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tracker/export/fieldsmapper/FieldsParamMapper.java:32]: Hey @lucaCambi77 ,

in my case, I am referring to the files under the dxf2 package. I would rather we don't change those, and rather create duplicate versions of them, to use in the new tracker module. That way we have less dependencies in dxf2 when we do remove the old tracker code from there.

As it stands, any work you do in the dxf2 package for master, will be removed in 2.41 most likely, if it relates to import at least, and probably to export as well, since the old endpoints for export will no longer be needed with the old importer gone.

[the remark this answers is not in the document]
Comment (lucaCambi77) [reply dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tracker/export/fieldsmapper/FieldsParamMapper.java:32]: The new tracker export is totally dependent on the dxf2 package see [here](https://github.com/dhis2/dhis2-core/blob/eb48b16bc219132cd877edec3428d402f78bdc59/dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tracker/export/TrackerEnrollmentsExportController.java#L44) for enrollments, here for [tracked entities](https://github.com/dhis2/dhis2-core/blob/6c3a535f28562b12bc7803ce8c102a9cbd86f982/dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tracker/export/TrackerTrackedEntitiesExportController.java#L54) . So i think first we should  think about how to deal with this dependency

**Comment by lucaCambi77** [comment]
https://github.com/dhis2/dhis2-core/pull/13236 has been created instead

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tracker/export/fieldsmapper/FieldsParamMapper.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (lucaCambi77) [reply dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tracker/export/fieldsmapper/FieldsParamMapper.java:32]: So what do we want to do with this PR? Shall we close it?

**Comment by lucaCambi77** [comment]
We close this PR as it seems the new tracker needs his own services. We keep it as a reference for future work if required
